# Analyzing borrower's risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customer's credit worthiness.

Your report will be considered when building a "credit scoring" of a potential customer. A "credit scoring" is used to evaluate the ability of a potential borrower to repay their loan.


**Description of the data:**
* children : the number of children in the family 
* days_employed : how long the customer has been working
* dob_years : the customer’s age
* education : the customer’s education level
* education_id : identifier for the customer’s education
* family_status : the customer’s marital status
* family_status_id : identifier for the customer’s marital status
* gender : the customer’s gender
* income_type : the customer’s income type
* debt : whether the customer has ever defaulted on a loan
* total_income : monthly income
* purpose : reason for taking out a loan

## Step 1. Open the data file and have a look at the general information. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Read the games data
credit_scoring_eng = pd.read_csv('/Users/fanibhushan/Downloads/credit_scoring_eng.csv')    
credit_scoring_eng.head()


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [3]:
#Number of rows & columns
'Total Rows & Columns:',credit_scoring_eng.shape

('Total Rows & Columns:', (21525, 12))

In [4]:
# variable types
credit_scoring_eng.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
credit_scoring_eng.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [6]:
# count of family_status having 20 children
credit_scoring_eng.loc[credit_scoring_eng["children"] == 20, "family_status"].value_counts()

married              49
civil partnership    12
unmarried             9
widow / widower       4
divorced              2
Name: family_status, dtype: int64

In [7]:
#count of family_status having 0 children
credit_scoring_eng.loc[credit_scoring_eng["children"] == 0, "family_status"].value_counts()

married              7500
civil partnership    2752
unmarried            2265
widow / widower       848
divorced              784
Name: family_status, dtype: int64

In [8]:
#Convert negative to positive
credit_scoring_eng['days_employed'] = abs(credit_scoring_eng['days_employed'])
credit_scoring_eng['children'] = abs(credit_scoring_eng['children'])
credit_scoring_eng['total_income'] = abs(credit_scoring_eng['total_income'])

In [9]:
credit_scoring_eng.describe(include=np.object)

education family_status gender income_type           purpose
count                 21525         21525  21525       21525             21525
unique                   15             5      3           8                38
top     secondary education       married      F    employee  wedding ceremony
freq                  13750         12380  14236       11119               797

In [10]:
#count of gender column
credit_scoring_eng['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [11]:
#Detecting problem entries
problem_entries = []

for row in credit_scoring_eng['gender']:
    if (row != 'F') & (row != 'M'):
        problem_entries.append(row)
print(problem_entries)

['XNA']


### Conclusion
There are total 21525 rows and 12 columns. Each columns consist of: int64,float64 & object datatype.

Descriptive statistics: shows "number of children in a family" max count is 20 & min count is -1 which is outlier. 

Family_status: Consist of 5 categories among them 57% is under married categories.
Income_type: Consist of 8 categories among them 51% is under emplyeed categories.
Also we get to know that gender column had 'XNA' value apart from male and female.

In 'credit_scoring_eng' dataset. How long the customers has been working & total income column both are in negative. That is not possible as number of days people working and their income cannot be in negative.So converted into positive. As negative numbers will impact in arithmatic calcultion.

## Step 2. Data preprocessing

### Preprocessing Missing Value

In [12]:
# col 'days_employed' & total_income round the decimal place to 2
credit_scoring_eng['days_employed'] = round(credit_scoring_eng['days_employed'],2)
credit_scoring_eng['total_income'] = round(credit_scoring_eng['total_income'],2)


In [13]:
#number of missing values per column
credit_scoring_eng.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [14]:
#creating a copy of orignal data
credit_scoring_eng_copy = credit_scoring_eng.copy()

In [15]:
#group data by gender and print mean & median of num_colums
num_columns = ['days_employed','total_income']
print('These are the means:')
print(credit_scoring_eng_copy.groupby('gender')[num_columns].mean())
print()
print('These are the medians:')
print(credit_scoring_eng_copy.groupby('gender')[num_columns].median())

These are the means:
        days_employed  total_income
gender                             
F        82397.770912  24655.604746
M        37000.331279  30907.144339
XNA       2358.600000  32624.820000

These are the medians:
        days_employed  total_income
gender                             
F            2539.855     21464.840
M            1662.370     26834.295
XNA          2358.600     32624.820


In [16]:
#group data & see the count of missing values in the two column
print('Count of missing days_employed by Gender')
print(credit_scoring_eng_copy[credit_scoring_eng_copy['days_employed'].isnull()]['gender'].value_counts())

print('Count of missing total_income by Gender')
print(credit_scoring_eng_copy[credit_scoring_eng_copy['total_income'].isnull()]['gender'].value_counts())

Count of missing days_employed by Gender
F    1484
M     690
Name: gender, dtype: int64
Count of missing total_income by Gender
F    1484
M     690
Name: gender, dtype: int64


In [17]:
#calculate & fill missing values for days_employed & total_income with mean & meadian
# Calculate values
Mediandaysemployed = credit_scoring_eng_copy['days_employed'].median()
Meantotalincome = credit_scoring_eng_copy['total_income'].mean()
Meandobyears = credit_scoring_eng_copy['dob_years'].mean()

credit_scoring_eng_copy.fillna(value= {'days_employed':Mediandaysemployed,'total_income':Meantotalincome,'dob_years':Meandobyears }, inplace = True)
credit_scoring_eng_copy.head(2)

children  days_employed  dob_years            education  education_id  \
0         1        8437.67         42    bachelor's degree             0   
1         1        4024.80         36  secondary education             1   

  family_status  family_status_id gender income_type  debt  total_income  \
0       married                 0      F    employee     0       40620.1   
1       married                 0      F    employee     0       17932.8   

                 purpose  
0  purchase of the house  
1           car purchase

In [18]:
# number of children in a family 20 replacing it with 2
credit_scoring_eng_copy.loc[credit_scoring_eng_copy['children'] == 20] = 2

In [19]:
#Check if there is any missing value
credit_scoring_eng_copy.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

#### Conclusion

"days_employed" & "total_income" column was float type so round the decimal place to 2.

Two Column has missing value i.e. 'days_imployed' & 'total_income'.  10% of row in "total_income" column has missing value and that is replaced by mean. In case of days_employed also have 10% of row missing and that is replaced by median.'dob_years' had age 0 which is not possible and so this was replaced by mean age.'no of children in family is' 20 & -1 which is also not possible, so converted -1 to 1 & 20 to 2

Reason for 'children' column: this can be typo error (instead of 1 it was -1 & instead of 2 it was 20) that could be the possible reason.

Reason for 'days_imployed', 'total_income' & 'dob_years' column: the vaues were missing because peple sometimes forget to mention or don't want to specify thier income, age & no of days employed. 

The means and medians are roughly equal for total_income & dob_years that means there are no outliers in this two columns so filled the missing value by mean.
For days_employed the mean is higher than the median that implies that it has outliers in this column so filled the missing vaue by median.


### Data type replacement

In [20]:
# checking the datatype
credit_scoring_eng_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [21]:
#transform from float to int type
credit_scoring_eng_copy['total_income']=credit_scoring_eng_copy['total_income'].astype('int64')
credit_scoring_eng_copy['days_employed'] = credit_scoring_eng_copy['days_employed'].astype('int64')

credit_scoring_eng_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


#### Conclusion

Converted  'total_income' & 'days_employed' to integer type. These columns were converted because float-point could be much slower than integer arithmetic. 

### Processing duplicates

In [22]:
#No of duplicate entries in dataset
'Duplicate entries in row:', credit_scoring_eng_copy.duplicated().sum()

('Duplicate entries in row:', 129)

In [23]:
credit_scoring_removed_duplicated = credit_scoring_eng.drop_duplicates()
len(credit_scoring_removed_duplicated)

21471

In [24]:
# drop XNA from gender column
credit_scoring_removed_duplicated = credit_scoring_removed_duplicated[(credit_scoring_removed_duplicated['gender'] == 'M') | (credit_scoring_removed_duplicated['gender'] == 'F')]
credit_scoring_removed_duplicated['gender'].value_counts()

F    14189
M     7281
Name: gender, dtype: int64

#### Conclusion

There were 129 rows duplicated. That means 0.59% of a variable is duplicated, delete the variable entirely.
Used drop_duplicates methods to repmove the duplicate as this method will return a new dataframe without the duplicates and then saved this into a new dataframe called 'credit_scoring_removed_duplicated'.

Possible reason for the appearence of duplicate data: Effective matching customers require multiple data points but if records does't have enough data points to match multiple customers may be associated with same records.

Dropped the row contaning 'XNA' in gender column.


### Categorizing Data

In [25]:
#Returns the category of income_level for the total_income column
def income_level(total_income):
    if (total_income > credit_scoring_removed_duplicated['total_income'].quantile(0.25)) & (total_income <= credit_scoring_eng['total_income'].quantile(0.50)):
        return 'medium'
    if total_income <= credit_scoring_removed_duplicated['total_income'].quantile(0.25):
        return 'low'
    if total_income <= credit_scoring_removed_duplicated['total_income'].quantile(0.75):
        return 'high'
    return 'very_high'

credit_scoring_removed_duplicated['income_level'] = credit_scoring_removed_duplicated['total_income'].apply(income_level)
credit_scoring_removed_duplicated.head()

children  days_employed  dob_years            education  education_id  \
0         1        8437.67         42    bachelor's degree             0   
1         1        4024.80         36  secondary education             1   
2         0        5623.42         33  Secondary Education             1   
3         3        4124.75         32  secondary education             1   
4         0      340266.07         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0      40620.10   
1            married                 0      F    employee     0      17932.80   
2            married                 0      M    employee     0      23341.75   
3            married                 0      M    employee     0      42820.57   
4  civil partnership                 1      F     retiree     0      25378.57   

                   purpose income_level  
0    purchase of the house    very_high  
1             car purchase       medium  
2    purchase of the house         high  
3  supplementary education    very_high  
4        to have a wedding         high

In [26]:
#Returns purpose_category for the purpose column
purpose= credit_scoring_removed_duplicated['purpose']


def get_purpose_category(purpose):
    if 'wedding' in purpose:
        return 'wedding'
    if 'house' in purpose:
        return 'house'
    if 'real estate' in purpose:
        return 'real estate'
    if 'property' in purpose:
        return 'property'
    if 'car' in purpose:
        return 'car'
    if 'cars' in purpose:
        return 'car'
    if 'education' in purpose:
        return 'education'
    if 'university' in purpose:
        return 'education'
    if 'educated' in purpose:
        return 'education'
    if 'housing' in purpose:
        return 'house'
    
    #return 'default'
    
credit_scoring_removed_duplicated['purpose_category'] = credit_scoring_removed_duplicated['purpose'].apply(get_purpose_category)
credit_scoring_removed_duplicated.head()

children  days_employed  dob_years            education  education_id  \
0         1        8437.67         42    bachelor's degree             0   
1         1        4024.80         36  secondary education             1   
2         0        5623.42         33  Secondary Education             1   
3         3        4124.75         32  secondary education             1   
4         0      340266.07         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0      40620.10   
1            married                 0      F    employee     0      17932.80   
2            married                 0      M    employee     0      23341.75   
3            married                 0      M    employee     0      42820.57   
4  civil partnership                 1      F     retiree     0      25378.57   

                   purpose income_level purpose_category  
0    purchase of the house    very_high            house  
1             car purchase       medium              car  
2    purchase of the house         high            house  
3  supplementary education    very_high        education  
4        to have a wedding         high          wedding

In [27]:
# mean of total_income categoried as per family_status & children
data_pivot = credit_scoring_removed_duplicated.pivot_table(index=['family_status','children'],columns='debt', values='total_income', aggfunc='mean')
data_pivot_reset_index = data_pivot.reset_index()
data_pivot_reset_index.head()

debt      family_status  children             0             1
0     civil partnership         0  26514.971544  25590.672376
1     civil partnership         1  27415.637237  25288.857843
2     civil partnership         2  27010.303584  27708.046538
3     civil partnership         3  29928.843077  18014.515000
4     civil partnership         4  26507.737143           NaN

In [28]:
# Convert the pivot table to dataframe for calculating avg_income_per_person
data_pivot_dataframe = pd.DataFrame(data_pivot_reset_index)

data_pivot_dataframe.loc[(data_pivot_dataframe['family_status'] == 'civil partnership') | (data_pivot_dataframe['family_status'] == 'married'),'avg_income_per_person'] = data_pivot_dataframe[1]/(data_pivot_dataframe['children'] +2)
data_pivot_dataframe.loc[(data_pivot_dataframe['family_status'] == 'divorced') | (data_pivot_dataframe['family_status'] == 'widow / widower')|(data_pivot_dataframe['family_status'] == 'unmarried'),'avg_income_per_person'] = data_pivot_dataframe[1]/(data_pivot_dataframe['children'] +1)
data_pivot_dataframe.head()

debt      family_status  children             0             1  \
0     civil partnership         0  26514.971544  25590.672376   
1     civil partnership         1  27415.637237  25288.857843   
2     civil partnership         2  27010.303584  27708.046538   
3     civil partnership         3  29928.843077  18014.515000   
4     civil partnership         4  26507.737143           NaN   

debt  avg_income_per_person  
0              12795.336188  
1               8429.619281  
2               6927.011635  
3               3602.903000  
4                       NaN

#### Conclusion
Categorised 'tota_income' column into low, medium , high & ver high income group.This was done on the basis of median and saved this into another column named 'income_level'.
Categorised 'purpose' column into different level of categories based on similar words from the purpose column.This data was saved into another column named 'purpose_category'.

By categorising the data created two columns named: 'income_level' & 'purpose_category'.
This was help us in futher analysis based on catagory.

Using pivot table: Aggregated the data & summarised it by grouping the 'family_status' with thier 'children'.'debt' provides an dditional way to segment the actual value we care. The column 'total_income' we see the average total income group wise.
From pivot table calculating average income per person. Using this average income per person will check the relationship if that particular category and repaying a loan has some relationship or not.

## Step 3. Answer these questions

### Is there a relation between having kids and repaying a loan on time?

In [29]:
credit_scoring_removed_duplicated['children'].value_counts()

0     14106
1      4856
2      2052
3       330
20       76
4        41
5         9
Name: children, dtype: int64

In [30]:
data_pivot_children = credit_scoring_removed_duplicated.pivot_table(index=['children','family_status'],columns='debt', values='total_income', aggfunc='mean')
data_pivot_with_reset_index_children = data_pivot_children.reset_index()
data_pivot_with_reset_index_children.tail(5)

debt  children      family_status             0             1
27          20  civil partnership  28298.033750  25238.483333
28          20           divorced           NaN  23127.120000
29          20            married  28525.750465  30712.220000
30          20          unmarried  16173.692000  18841.370000
31          20    widow / widower  21667.800000           NaN

In [31]:
data_pivot_dataframe_per_children= data_pivot_dataframe.pivot_table(index= ['children'] ,values ='avg_income_per_person', aggfunc='mean')
data_pivot_dataframe_per_children['avg_income_per_person'].sort_values(ascending= False)

children
0     20273.513706
1     11696.098171
2      8255.350442
3      5219.697143
4      4100.149250
20     1135.428328
Name: avg_income_per_person, dtype: float64

In [32]:
def children_impact(children):
    if children <= data_pivot_dataframe_per_children['avg_income_per_person'].median():
        return 'high_risk'
    else:
        return 'low_risk'
    
data_pivot_dataframe_per_children['children_impact'] = data_pivot_dataframe_per_children['avg_income_per_person'].apply(children_impact) 
data_pivot_dataframe_per_children

debt      avg_income_per_person children_impact
children                                       
0                  20273.513706        low_risk
1                  11696.098171        low_risk
2                   8255.350442        low_risk
3                   5219.697143       high_risk
4                   4100.149250       high_risk
20                  1135.428328       high_risk

#### Conclusion

Number of Children family has is: 0,1,2,3,4

From pivot table we can see that on the basis children & their family_status for both when person is under debt & no debt what is the average total income.
As we are looking for repayment of loan on time: So we will consider person under debt that means where debt==1 & what is the average income per person categorised by family_status & children.
Further when we see on the basis of number of children family has and average income per person. Categorised the impact children can have based on average income per person by taking the median and classified under high & low risk.

From the above table we can say based: Greater the average income per person with less number of children there is higher chances of repayment of loan on time.

As we can see family having more than 2 children their aberage imcome per person decreases and hence, repayment of loan timely is under high risk. 

Thus, we can conclude that yes there is relationship between having kids and repayment of loan on time.

### Is there a relation between marital status and repaying a loan on time?

In [33]:
credit_scoring_removed_duplicated['family_status'].value_counts()

married              12344
civil partnership     4162
unmarried             2810
divorced              1195
widow / widower        959
Name: family_status, dtype: int64

In [34]:
data_pivot_family_status = credit_scoring_removed_duplicated.pivot_table(index=['family_status','children'],columns='debt', values='total_income', aggfunc='mean')
data_pivot_with_reset_index_family_status = data_pivot_family_status.reset_index()

In [35]:
data_pivot_dataframe_per_family_status= data_pivot_dataframe.pivot_table(index= ['family_status'] ,values ='avg_income_per_person', aggfunc='mean')
data_pivot_dataframe_per_family_status['avg_income_per_person'].sort_values(ascending= False)

family_status
widow / widower      16804.228542
divorced             10340.491754
unmarried             9801.300503
married               6820.098445
civil partnership     6580.414778
Name: avg_income_per_person, dtype: float64

In [36]:
def family_status_impact(family_status):
    if family_status < data_pivot_dataframe_per_family_status['avg_income_per_person'].median():
        return 'high_risk'
    else:
        return 'low_risk'
    
data_pivot_dataframe_per_family_status['family_status_impact'] = data_pivot_dataframe_per_family_status['avg_income_per_person'].apply(family_status_impact) 
data_pivot_dataframe_per_family_status

debt               avg_income_per_person family_status_impact
family_status                                                
civil partnership            6580.414778            high_risk
divorced                    10340.491754             low_risk
married                      6820.098445            high_risk
unmarried                    9801.300503             low_risk
widow / widower             16804.228542             low_risk

#### Conclusion

Family Status is divided into 4 categories: married,civil partnership,unmarried,divorced,widow / widower

From the above table we can say: Divorced, unmarried and widow there is higher chances of repayment of loan on time.

As we can see family status for civil partner, married imcome per person is lower and hence, repayment of loan timely is under high risk. 

Thus, we can conclude that there is relationship between family_status and repayment of loan on time.

### Is there a relation between income level and repaying a loan on time?

In [37]:
credit_scoring_removed_duplicated['income_level'].value_counts()

very_high    6958
medium       4838
low          4838
high         4836
Name: income_level, dtype: int64

In [38]:
data_pivot_income_level = credit_scoring_removed_duplicated.pivot_table(index=['income_level','family_status','children'],columns='debt', values='total_income', aggfunc='mean')
data_pivot_reset_index_income_level = data_pivot_income_level.reset_index()
data_pivot_reset_index_income_level.tail(5)

debt income_level    family_status  children             0         1
101     very_high  widow / widower         0  44668.677417  40388.46
102     very_high  widow / widower         1  41397.720667  45850.67
103     very_high  widow / widower         2  41229.720000  51284.32
104     very_high  widow / widower         3  44415.853333       NaN
105     very_high  widow / widower        20  42315.970000       NaN

In [39]:
data_pivot_dataframe_income_level = pd.DataFrame(data_pivot_reset_index_income_level)

data_pivot_dataframe_income_level.loc[(data_pivot_dataframe_income_level['family_status'] == 'civil partnership') | (data_pivot_dataframe_income_level['family_status'] == 'married'),'avg_income_per_person'] = data_pivot_dataframe_income_level[1]/(data_pivot_dataframe_income_level['children'] +2)
data_pivot_dataframe_income_level.loc[(data_pivot_dataframe_income_level['family_status'] == 'divorced') | (data_pivot_dataframe_income_level['family_status'] == 'widow / widower')|(data_pivot_dataframe_income_level['family_status'] == 'unmarried'),'avg_income_per_person'] = data_pivot_dataframe_income_level[1]/(data_pivot_dataframe_income_level['children'] +1)


In [40]:
data_pivot_dataframe_per_income_level= data_pivot_dataframe_income_level.pivot_table(index= ['income_level'] ,values ='avg_income_per_person', aggfunc='mean')
data_pivot_dataframe_per_income_level['avg_income_per_person'].sort_values(ascending= False)

income_level
very_high    19648.215039
high         11848.302133
medium        7962.952118
low           6044.830359
Name: avg_income_per_person, dtype: float64

In [41]:
def income_level_impact(income_level):
    if income_level <= data_pivot_dataframe_per_income_level['avg_income_per_person'].median():
        return 'high_risk'
    else:
        return 'low_risk'
    
data_pivot_dataframe_per_income_level['income_level_impact'] = data_pivot_dataframe_per_income_level['avg_income_per_person'].apply(income_level_impact) 
data_pivot_dataframe_per_income_level

debt          avg_income_per_person income_level_impact
income_level                                           
high                   11848.302133            low_risk
low                     6044.830359           high_risk
medium                  7962.952118           high_risk
very_high              19648.215039            low_risk

#### Conclusion

Income level is divided into 4 categories: low,meduim ,high,very high

From the above table we can say: those who are under high and very high income group there is higher chances of repayment of loan on time.

As we can see income level for low and medium imcome group has lower income per person and hence, repayment of loan timely is under high risk. 

Thus, we can conclude that there is relationship between income_level and repayment of loan on time.


### How do different loan purposes affect on-time repayment of the loan?

In [42]:
credit_scoring_removed_duplicated['purpose_category'].value_counts()

real estate    4465
car            4308
education      4014
house          3809
property       2539
wedding        2335
Name: purpose_category, dtype: int64

In [43]:
data_pivot_purpose_category = credit_scoring_removed_duplicated.pivot_table(index=['purpose_category','family_status','children'],columns='debt', values='total_income', aggfunc='mean')
data_pivot_reset_index_purpose_category = data_pivot_purpose_category.reset_index()

In [44]:
data_pivot_dataframe_purpose_category = pd.DataFrame(data_pivot_reset_index_purpose_category)


data_pivot_reset_index_purpose_category.loc[(data_pivot_reset_index_purpose_category['family_status'] == 'civil partnership') | (data_pivot_reset_index_purpose_category['family_status'] == 'married'),'avg_income_per_person'] = data_pivot_reset_index_purpose_category[1]/(data_pivot_reset_index_purpose_category['children'] +2)
data_pivot_reset_index_purpose_category.loc[(data_pivot_reset_index_purpose_category['family_status'] == 'divorced') | (data_pivot_reset_index_purpose_category['family_status'] == 'widow / widower')|(data_pivot_reset_index_purpose_category['family_status'] == 'unmarried'),'avg_income_per_person'] = data_pivot_reset_index_purpose_category[1]/(data_pivot_reset_index_purpose_category['children'] +1)

data_pivot_reset_index_purpose_category.head(5)

debt purpose_category      family_status  children             0  \
0                 car  civil partnership         0  26347.409106   
1                 car  civil partnership         1  27616.252530   
2                 car  civil partnership         2  26815.049643   
3                 car  civil partnership         3  21212.050000   
4                 car  civil partnership         4  20140.810000   

debt             1  avg_income_per_person  
0     22871.033600           11435.516800  
1     25776.758125            8592.252708  
2     22454.403333            5613.600833  
3     42181.290000            8436.258000  
4              NaN                    NaN

In [45]:
data_pivot_dataframe_per_purpose_category= data_pivot_dataframe_purpose_category.pivot_table(index= ['purpose_category'] ,values ='avg_income_per_person', aggfunc='mean')
data_pivot_dataframe_per_purpose_category['avg_income_per_person'].sort_values(ascending= False)

purpose_category
house          12439.404242
property       12393.080100
real estate    12198.124821
education      11097.474009
car            10987.309812
wedding         6000.334237
Name: avg_income_per_person, dtype: float64

In [46]:
def purpose_category_impact(purpose_category):
    if purpose_category <= data_pivot_dataframe_per_family_status['avg_income_per_person'].median():
        return 'high_risk'
    else:
        return 'low_risk'
    
data_pivot_dataframe_per_purpose_category['purpose_category_impact'] = data_pivot_dataframe_per_purpose_category['avg_income_per_person'].apply(purpose_category_impact) 
data_pivot_dataframe_per_purpose_category

debt              avg_income_per_person purpose_category_impact
purpose_category                                               
car                        10987.309812                low_risk
education                  11097.474009                low_risk
house                      12439.404242                low_risk
property                   12393.080100                low_risk
real estate                12198.124821                low_risk
wedding                     6000.334237               high_risk

#### Conclusion

Purpose of loan is divided into various categories: car, education ,house, property, real estate,wedding

From the above table we can say: those who take loan for education, house, property & real estate there is higher chances of repayment of loan on time.

As we can see those who take loan on car and for wedding purpose, repayment of loan timely is under high risk. 

Thus, we can conclude that there is relationship between purpose effect and repayment of loan on time.

## Step 4. General Conclusion

Fom the above analysis we can conclude that repayment of loan on time depends on several facrtors such as: income of the family, number of children in the family, marital status of the family and purpose of their loan.
This is concluded on the fact that if we calculate average income per person of the marital status of the family we see the result shows married & civil patner has lower average income per person as compared to other categories of family status. In case of number of children and average income per person family having more than 1 children their average income per person is lower. Income level with low and medium income group in that case also average income per person is lower. Finally with purpose for loan those who has lower average income per person takes loan for car and wedding.

To conclude: When average per person income is lower then, family status:married & civil partner and number of children: greater than 1 children and income_level: lower & medium income group and purpose: loan for car and wedding. High chances that they won't pay loan on-time. Hence, we can say all the above factors affect repayment of loan on-time.